TesseractOCR/G8Tesseract.mm

```diff
@@ -310,18 +340,124 @@ - (void)setVariablesFromDictionary:(NSDictionary *)dictionary
+    
+    self.imageSize = image.size; //self.imageSize used in the characterBoxes method
+    
+    if (self.isEngineConfigured) {
+        Pix *pix = nullptr;
+        
+        if ([self.delegate respondsToSelector:@selector(preprocessedImageForTesseract:sourceImage:)]) {
+            UIImage *thresholdedImage = [self.delegate preprocessedImageForTesseract:self sourceImage:image];
+            if (thresholdedImage != nil) {
+                self.imageSize = thresholdedImage.size;
+                
+                Pix *pixs = [self pixForImage:thresholdedImage];
+                pix = pixConvertTo1(pixs, UINT8_MAX / 2);
+                pixDestroy(&pixs);
+                
+                if (pix == nullptr) {
+                    NSLog(@"WARNING: Can't create Pix for custom thresholded image!");
+                }
+            }
+        }
+        
+        if (pix == nullptr) {
+            pix = [self pixForImage:image];
+        }
+        
+        @try {
+            _tesseract->SetImage(pix);
+        }
+        //LCOV_EXCL_START
+        @catch (NSException *exception) {
+            NSLog(@"ERROR: Can't set image: %@", exception);
+        }
+        //LCOV_EXCL_STOP
+        pixDestroy(&pix);
+    }
+    
+    _image = image;
+
+    [self resetFlags];
+}
+
+- (void)setEngineSourceResolution:(NSUInteger)sourceResolution {
+    
+    if (self.isEngineConfigured) {
+        _tesseract->SetSourceResolution((int)sourceResolution);
+    }
+}
+
+- (void)setEngineRect:(CGRect)rect {
+    
+    if (!self.isEngineConfigured) {
+        return;
+    }
+    
+    CGFloat x = CGRectGetMinX(rect);
+    CGFloat y = CGRectGetMinY(rect);
+    CGFloat width = CGRectGetWidth(rect);
+    CGFloat height = CGRectGetHeight(rect);
+    
+    // Because of custom preprocessing we may have to resize rect
+    if (CGSizeEqualToSize(self.image.size, self.imageSize) == NO) {
+        CGFloat widthFactor = self.imageSize.width / self.image.size.width;
+        CGFloat heightFactor = self.imageSize.height / self.image.size.height;
+        
+        x *= widthFactor;
+        y *= heightFactor;
+        width *= widthFactor;
+        heightFactor *= heightFactor;
+    }
+    
+    CGFloat (^clip)(CGFloat, CGFloat, CGFloat) = ^(CGFloat value, CGFloat min, CGFloat max) {
+        return (value < min ? min : (value > max ? max : value));
+    };
+    
+    // Clip rect by image size
+    x = clip(x, 0, self.imageSize.width);
+    y = clip(y, 0, self.imageSize.height);
+    width = clip(width, 0, self.imageSize.width - x);
+    height = clip(height, 0, self.imageSize.height - y);
+    
+    _tesseract->SetRectangle(x, y, width, height);
 }
 
-#pragma mark - Getters and setters
+#pragma mark - Public getters and setters
 
 - (void)setLanguage:(NSString *)language
 {
-    if ([_language isEqualToString:language] == NO) {
-        _language = [language copy];
-
+    if ([language isEqualToString:_language] == NO || (!language && _language) ) {
```

Comment:
That condition is necessary to eliminate setter run if we try to set nil to the already empty property.
Without this extra condition the condition `[language isEqualToString:_language] == NO` or `[language isEqualToString:_language] == NO` are true if both `language == nil` and `_language == nil`. That's a very rear case, I know. but we have it somewhere in our tests, so I've added it here as well.